app/lib/search/index_simple.dart

```diff
@@ -227,29 +231,70 @@ class SimplePackageIndex implements PackageIndex {
 }
 
 class Score {
-  final Map<String, double> values = <String, double>{};
+  final Map<String, double> _values;
 
-  Iterable<String> getKeys() => values.keys;
+  Score(Map<String, double> values) : _values = new Map.unmodifiable(values);
 
-  void addValues(Map<String, double> newValues, double weight) {
-    if (newValues == null) return;
-    newValues.forEach((String key, double score) {
-      if (score != null) {
-        final double prev = values[key] ?? 0.0;
-        values[key] = prev + score * weight;
-      }
-    });
+  Set<String> getKeys() => _values.keys.toSet();
+  double getMaxValue() => _values.values.fold(0.0, max);
+  Map<String, double> getValues() => _values;
+
+  double operator [](String key) => _values[key] ?? 0.0;
+
+  /// Calculates the intersection of the two Scores, by multiplying the values.
+  ///
+  /// [weight] can be used to dampen the removal-effect of multiplication with
+  /// zero or very low values, and to keep results from the original set.
+  Score multiply(Score other, {double weight: 0.0}) {
+    final Set<String> keys = getKeys().intersection(other.getKeys());
+    return new Score(new Map.fromIterable(
+      keys,
+      value: (key) {
+        final double otherValue = other[key];
+        final double m = otherValue * weight + (1.0 - weight);
+        return this[key] * m;
+      },
+    ));
+  }
+
+  /// Calculates the union of the two Scores, by using the maximum values from
+  /// the sets.
+  Score union(Score other) {
+    final Set<String> keys = getKeys()..addAll(other.getKeys());
+    final result = <String, double>{};
+    for (String key in keys) {
+      result[key] = max(this[key], other[key]);
+    }
+    return new Score(result);
   }
 
-  void removeWhere(bool keyCondition(String key)) {
-    final Set<String> keysToRemove = values.keys.where(keyCondition).toSet();
-    keysToRemove.forEach(values.remove);
+  Score removeLowValues({double fraction, double minValue}) {
+    double threshold = minValue;
+    if (fraction != null) {
+      final double fractionValue = getMaxValue() * fraction;
+      threshold ??= fractionValue;
+      threshold = max(threshold, fractionValue);
+    }
+    if (threshold == null) {
+      return this;
+    }
+    final result = <String, double>{};
+    for (String key in _values.keys) {
+      final double value = _values[key];
+      if (value < threshold) continue;
+      result[key] = value;
+    }
+    return new Score(result);
   }
 
-  void removeLowScores(double fraction) {
-    final double maxValue = values.values.fold(0.0, max);
-    final double cutoff = maxValue * fraction;
-    removeWhere((key) => values[key] < cutoff);
+  Score withKeys(Iterable<String> keys) {
```

Comment:
Maybe `restrictToKeys` or `project` (from [relational database terminology](https://en.wikipedia.org/wiki/Projection_(relational_algebra))) would be better.